apps/src/lib/node/ledger/events/log.rs

```diff
@@ -0,0 +1,893 @@
+        {
+            ::tokio::task::block_in_place(|| $expr)
+        }
+    }};
+}
+
+/// Parameters to configure the pruning of the event log.
+#[derive(Debug, Copy, Clone)]
+pub struct Params {
+    /// Soft lock on the maximum number of events the event log can hold.
+    ///
+    /// If the number of events in the log exceeds this value, the log
+    /// will be pruned.
+    pub max_log_events: usize,
+    /// Soft lock on the number of entries the event log can hold.
+    ///
+    /// If the difference between the newest log entry and the oldest's
+    /// block heights is greater than this value, the log will be pruned.
+    pub log_block_height_diff: u64,
+}
+
+impl Default for Params {
+    fn default() -> Self {
+        // TODO: tune the default params
+        Self {
+            max_log_events: 50000,
+            log_block_height_diff: 1000,
+        }
+    }
+}
+
+/// Instantiates a new event log and its associated machinery.
+///
+/// General usage flow:
+///
+///   1. Spawn a new asynchronous task, with a [`Logger`]
+///      running on a loop.
+///   2. Send new events to the [`Logger`] with a [`LogEntrySender`].
+///      This will alter the state of the [`EventLog`].
+///   3. Concurrently, other asynchronous tasks may access the
+///      [`EventLog`] to check for new events.
+pub fn new(params: Params) -> (EventLog, Logger, LogEntrySender) {
+    let (tx, rx) = mpsc::unbounded_channel();
+
+    let log = EventLog::new(params);
+    let logger = Logger {
+        receiver: rx,
+        log: log.clone(),
+    };
+    let sender = LogEntrySender { sender: tx };
+
+    (log, logger, sender)
+}
+
+/// Represents an entry in the event log.
+#[derive(Debug, Clone)]
+pub struct LogEntry {
+    /// The block height at which we emitted the events.
+    pub block_height: BlockHeight,
+    /// The events emitted by a `FinalizeBlock` call.
+    pub events: Vec<Event>,
+}
+
+/// Represents a node in the linked list of log entries.
+#[derive(Debug)]
+struct LogNode {
+    entry: LogEntry,
+    next: Option<Arc<LogNode>>,
+}
+
+impl LogNode {
+    /// Return an iterator over the given linked list
+    /// of [`LogNode`] instances.
+    fn iter(node: Option<&Arc<LogNode>>) -> LogNodeIter<'_> {
+        LogNodeIter { node }
+    }
+}
+
+/// Iterator over [`LogNode`] instances in
+/// the same linked list.
+#[derive(Debug)]
+struct LogNodeIter<'a> {
+    node: Option<&'a Arc<LogNode>>,
+}
+
+impl<'a> Iterator for LogNodeIter<'a> {
+    type Item = &'a Arc<LogNode>;
+
+    fn next(&mut self) -> Option<&'a Arc<LogNode>> {
+        self.node.take().map(|node| {
+            self.node = node.next.as_ref();
+            node
+        })
+    }
+}
+
+/// Represents a log of [`Event`] instances emitted by
+/// `FinalizeBlock` calls, in the ledger.
+///
+/// __INVARIANT:__ All logged events should be ordered
+/// in reverse by the shell's block height at the time
+/// of calling `FinalizeBlock`.
+#[derive(Debug, Clone)]
+pub struct EventLog {
```

Comment:
It might tighten up the code to implement `Deref` and `DerefMut` for this type

Reply:
It essentially is just a pointer.

Reply:
unfortunately I can't implement `Deref`, because `EventLogInner` is a private type